docs/architecture/ADR-009-liquidity-mining.md

```diff
@@ -0,0 +1,268 @@
+
+### Incentive Programs
+
+The `x/incentive` module exists to incentivize users to lock collateral assets at the various tiers.
+
+The basic unit of incentivization is the `IncentiveProgram`, which, continuously between a start time and end time, distributes a fixed amount of a given asset type across all holders of a given denom of locked `uToken` collateral, weighted by amount held and locking tier.
+
+```go
+type IncentiveProgram struct {
+  ID               uint32
+  LockedDenom      string
+  RewardDenom      string
+  StartTime        uint64
+  Duration         uint64
+}
+```
+
+Start time is unix time, and duration is measured in seconds.
+
+TotalRewards is the total amount of RewardDenom to be distributed as a result of the incentive program.
+
+Incentive programs, both active and historical, will be stored in state using protobuf marshaling and a simple uint32 ID once their initial governance proposal passes. Their total rewards, which are set when a valid program is funded, will also be stored by program ID.
+
+```go
+IncentiveProgramPrefix | ID => IncentiveProgram
+ProgramRewardsPrefix   | ID => sdk.Coin
+```
+
+### Governance
+
+A single governance proposal type is needed to create new incentive programs.
+
+```go
+type IncentiveProgramProposal struct {
+  Title       string
+  Description string
+  Program     IncentiveProgram
+}
+```
+
+Note: After Cosmos SDK 0.46, this can be simplified to use a concrete message type only usabel by governance.
+
+See _Funding Proposals_ for additional considerations.
+
+### Reward Math
+
+Our main requirement is to continuously distribute rewards without iterating over addresses.
+
+The following approach is proposed:
+
+- For every user, each nonzero `Locked(address,denom,tier) = Amount` is stored.
+- Additionally, each nonzero `TotalLocked(denom,tier) = Amount` is kept up to date in state. These track the sum of uTokens locked across all addresses.
+- For each `(denom,tier)` that has ever been incentivized, any nonzero `RewardAccumulator(denom,tier)` is stored in state. This `sdk.DecCoins` represents the total rewards a single locked `uToken` would have accumulated if locked into a given tier at genesis.
+- For each user, any nonzero `PendingReward` is stored as `sdk.DecCoins`.
+- For each nonzero `Locked(address,denom,tier)`, a value `RewardBasis(address,denom,tier)` is stored which tracks the value of `RewardAccumulator(denom,tier)` at the most recent time rewards were added to `PendingReward` by the given (address,denom,tier).
+- At any given `EndBlock`, each active incentive program performs some computations:
+  - Calculates the total `RewardDenom` rewards that will be given by the program in the current block `X = program.TotalRewards * (secondsElapsed / program.Duration)`
```

Comment:
Are secondsElapsed calculated using block time & block height like in the oracle module? If not, if the chain goes down, will time continue to pass?

Reply:
Block time, yeah. Leverage module does the same with interest too